src/migrate.c

```diff
@@ -0,0 +1,383 @@
+        }
+    }
+
+    LOG_VERBOSE("didn't find shardgroup slot range for this");
+    return false;
+}
+
+static bool validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return true;
+    }
+
+    return false;
+}
+
+static bool validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return true;
+    }
+
+    return false;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    int ret = RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len);
+    RedisModule_Assert(ret == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    unsigned int slot = keyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMigrationSessionKey(rr, slot, import_keys.migration_session_key)) {
+        LOG_DEBUG("ignoring import keys as incorrect migration session key");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid migration_session_key");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        /* this is expected situation (suspended old leader) */
+        LOG_DEBUG("ignoring import keys as old term");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid term");
+        }
+        goto exit;
+    }
+
+    /* 2  'static' strings need for restore command */
+    RedisModuleString *zero = RedisModule_CreateString(rr->ctx, "0", strlen("0")); /* ttl of zero */
+    RedisModuleString *replace = RedisModule_CreateString(rr->ctx, "REPLACE", strlen("REPLACE")); /* overwrite on import */
+
+    for (size_t i = 0; i < import_keys.num_keys; i++) {
+        RedisModuleString *temp[4];
+        temp[0] = import_keys.key_names[i];
+        temp[1] = zero;
+        temp[2] = import_keys.key_serialized[i];
+        temp[3] = replace;
+
+        enterRedisModuleCall();
+        RedisModuleCallReply *reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 4);
+        exitRedisModuleCall();
+        RedisModule_Assert(reply != NULL);
+
+        if (RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ERROR) {
+            size_t err_len;
+            const char *err = RedisModule_CallReplyStringPtr(reply, &err_len);
+            LOG_WARNING("importKeys: restore failed with %.*s", (int) err_len, err);
+            RedisModule_Assert(RedisModule_CallReplyType(reply) != REDISMODULE_REPLY_ERROR);
```

Comment:
This will be true all the time. Do we need this if check rather than just an assert?